### PR TITLE
Add SubNetwork struct to allow passing tree structure to AutoHOOT

### PR DESCRIPTION
**Pull request by LinjianMa**
The goal of this PR is to allow passing tree structures to AutoHOOT.

Currently, the batch_tensor_contraction function takes arrays of tensors as inputs, for instance
`batch_tensor_contraction([network1, network2])`, where `network1=[A,B,C,D]`, `network2=[A,B,C,D,E]`. The disadvantage is that AutoHOOT needs to figure out what's the optimal contraction path for each network, and when the network size is large, it becomes difficult to find an efficient path.

The goal is to take tree structure as input, so that AutoHOOT only needs to find optimal paths for each sub-contraction. For instance, when we call `batch_tensor_contraction([tree1, tree2])`, and tree1 includes a structure equivalent to `[[A,B],[C,D,E]]`, then AutoHOOT only needs to find paths for the subnetwork [A,B] and [C,D,E].

This API is prepared for more efficient PEPS calculations.

The tree structured is represented by the struct ContractNode. ContractNode(A, ContractNode(B, C, D)) represents the contraction path [A, [B, C, D]].

### CODE REVIEW

File: src/ITensorAutoHOOT/batch_contraction.jl [hunk 1]
```diff
@@ -42,18 +52,55 @@ function Executor(networks::Array, cache::NetworkCache)
...
+end
+
+@non_differentiable NetworkCache(trees::Array{ContractNode,1})
+
+"""
+Generate a cached network, under the constraint that tensors in contract_order will be
+contracted based on the order from Array start to the end.
+"""
+# NOTE: this function is experimental and it hasn't been used in experiments yet.
+function NetworkCache(networks::Array{Array{ITensor,1},1}, contract_order::Array{ITensor})
```
Comment (LinjianMa) [inline src/ITensorAutoHOOT/batch_contraction.jl:88]: This function is not used in any experiments. It's just used for me to test the efficiency of the constrained contraction path algorithm. Unfortunately, it didn't perform very well.

File: src/ITensorAutoHOOT/batch_contraction.jl [hunk 2]
```diff
@@ -25,15 +27,23 @@ end
 
 NetworkSum() = NetworkSum([])
 
-function Executor(networks::Array)
-  nodes, node_dict = generate_einsum_expr(networks)
-  net_sums = [NetworkSum([go.generate_optimal_tree(n)]) for n in nodes]
+function Executor(networks::Array{Union{Array{ITensor,1}},1})
```
Comment (mtfishman) [inline src/ITensorAutoHOOT/batch_contraction.jl:30]: Can use `Vector{ITensor}` instead of `Array{ITensor,1}` (it is just an alias for that).

**Comment by mtfishman** [comment]
Gotcha, so `ContractNode` wraps a part of the subnetwork that needs to get optimized, but if part of the network is not wrapped in `ContractNode` then it is not optimized? So for example:
```julia
[ContractNode([A, B, C]), ContractNode([D, E, F]), ContractNode([G, H, I])]`
```
means the subnetworks `[A, B, C]`, `[D, E, F]`, and `[G, H, I]` will get optimized, but then the resulting network of those contractions will not be optimized?

**Comment by mtfishman** [comment]
What if we called it `SubNetwork`?

**Comment by LinjianMa** [comment]
> What if we called it `SubNetwork`?

sounds good

**Comment by LinjianMa** [comment]
> Gotcha, so `ContractNode` wraps a part of the subnetwork that needs to get optimized, but if part of the network is not wrapped in `ContractNode` then it is not optimized? So for example:
> 
> ```julia
> [ContractNode([A, B, C]), ContractNode([D, E, F]), ContractNode([G, H, I])]
> ```
> 
> means the subnetworks `[A, B, C]`, `[D, E, F]`, and `[G, H, I]` will get optimized, but then the resulting network of those contractions will not be optimized?

Yes, so each ContractNode represents a network, and for your given expression, we can think of it as a batch of ContractNodes, so we will not optimize them.

**Comment by mtfishman** [comment]
Ok, that makes sense. This will be very useful to have.

One design question I have is, should `[A, B, C]` optimize by default, our should `SubNetwork([A, B, C])`? Maybe the name should convey whether or not it will optimize. I could imagine either:
```julia
[OptNetwork([A, B, C]), D, E]
```
where for `ABC = OptNetwork([A, B, C])`, `[ABC, D, E]` does not get optimized, or
```julia
NoOptNetwork([[A, B, C], D, E])
```
where `[A, B, C]` gets optimized by default but then `[ABC, D, E]` does not get optimized. Maybe it is better the way you have it with not optimizing by default, since optimizing can have a heavy cost so should be opt-in.

Maybe the name `OptNetwork` is better so it indicates what it's purpose is.

Over all this looks good to me, I will have to play around with it a bit to understand the details and interface better but if it helps with the next steps of the project then go ahead and merge.

**Comment by LinjianMa** [comment]
OK I see where the confusion is. In my design, the ContractNode struct represents a network, and the array does not. So what 
[node1, node2, node3] represents is not a network, but three different ones. 

Therefore, for now we don't allow the representation such as `ABCDE = ContractNode([[A, B, C], D, E])`, one should write it as `ABCDE = ContractNode(ContractNode(A, B, C), D, E)`. 

We can choose to generate AutoHOOT expressions for it, either with the optimized contraction path or not. To do that we can call
`generate_einsum_expr([ABCDE], optimize=false)`, which will generate something like
`einsum(str, einsum(str, A, B, C), D, E)`, or
`generate_einsum_expr([ABCDE], optimize=true)`, which will also optimize both einsums `einsum(str, A, B, C)` and `einsum(str, ABC, D, E)`.

**Comment by mtfishman** [comment]
Gotcha, thanks for the clarification. So I think in that case `SubNetwork` is a good name.

**Review by mtfishman — APPROVED** [review]

**Pull request merged**